// ocaml-language-server/bin/server/command/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const getFormatted = require("./getFormatted");
exports.getFormatted = getFormatted;
const getAvailableLibraries_1 = require("./getAvailableLibraries");
exports.getAvailableLibraries = getAvailableLibraries_1.default;
const getDocumentation_1 = require("./getDocumentation");
exports.getDocumentation = getDocumentation_1.default;
const getMerlinFiles_1 = require("./getMerlinFiles");
exports.getMerlinFiles = getMerlinFiles_1.default;
const getModules_1 = require("./getModules");
exports.getModules = getModules_1.default;
const getOccurrences_1 = require("./getOccurrences");
exports.getOccurrences = getOccurrences_1.default;
const getPrefix_1 = require("./getPrefix");
exports.getPrefix = getPrefix_1.default;
const getProjectEnv_1 = require("./getProjectEnv");
exports.getProjectEnv = getProjectEnv_1.default;
const getText_1 = require("./getText");
exports.getText = getText_1.default;
const getTextDocument_1 = require("./getTextDocument");
exports.getTextDocument = getTextDocument_1.default;
const getType_1 = require("./getType");
exports.getType = getType_1.default;
const getWordAtPosition_1 = require("./getWordAtPosition");
exports.getWordAtPosition = getWordAtPosition_1.default;
const restartMerlin_1 = require("./restartMerlin");
exports.restartMerlin = restartMerlin_1.default;
