// ocaml-language-server/bin/server/feature/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codeAction_1 = require("./codeAction");
exports.codeAction = codeAction_1.default;
const codeLens_1 = require("./codeLens");
exports.codeLens = codeLens_1.default;
const codeLensResolve_1 = require("./codeLensResolve");
exports.codeLensResolve = codeLensResolve_1.default;
const completion_1 = require("./completion");
exports.completion = completion_1.default;
const completionResolve_1 = require("./completionResolve");
exports.completionResolve = completionResolve_1.default;
const definition_1 = require("./definition");
exports.definition = definition_1.default;
const didChangeConfiguration_1 = require("./didChangeConfiguration");
exports.didChangeConfiguration = didChangeConfiguration_1.default;
const didChangeWatchedFiles_1 = require("./didChangeWatchedFiles");
exports.didChangeWatchedFiles = didChangeWatchedFiles_1.default;
const documentFormatting_1 = require("./documentFormatting");
exports.documentFormatting = documentFormatting_1.default;
const documentHighlight_1 = require("./documentHighlight");
exports.documentHighlight = documentHighlight_1.default;
const documentSymbol_1 = require("./documentSymbol");
exports.documentSymbol = documentSymbol_1.default;
const hover_1 = require("./hover");
exports.hover = hover_1.default;
const references_1 = require("./references");
exports.references = references_1.default;
const rename_1 = require("./rename");
exports.rename = rename_1.default;
const workspaceSymbol_1 = require("./workspaceSymbol");
exports.workspaceSymbol = workspaceSymbol_1.default;
